src/data/shortlist.cpp

```diff
@@ -0,0 +1,396 @@
+    bool check = vals.size() > 1 ? std::stoi(vals[1]) : 1;
+    LOG(info, "[data] Loading binary shortlist as {} {}", fname, check);
+    load(fname, check);
+  }
+  else{
+    firstNum_ = vals.size() > 1 ? std::stoi(vals[1]) : 100;
+    bestNum_ = vals.size() > 2 ? std::stoi(vals[2]) : 100;
+    float threshold = vals.size() > 3 ? std::stof(vals[3]) : 0;
+    LOG(info, "[data] Importing text lexical shortlist as {} {} {} {}",
+        fname, firstNum_, bestNum_, threshold);
+    import(fname, threshold);
+  }
+}
+
+Ptr<Shortlist> BinaryShortlistGenerator::generate(Ptr<data::CorpusBatch> batch) const {
+  auto srcBatch = (*batch)[srcIdx_];
+  size_t srcVocabSize = srcVocab_->size();
+  size_t trgVocabSize = trgVocab_->size();
+
+  // Since V=trgVocab_->size() is not large, anchor the time and space complexity to O(V).
+  // Attempt to squeeze the truth tables into CPU cache
+  std::vector<bool> srcTruthTable(srcVocabSize, 0);  // holds selected source words
+  std::vector<bool> trgTruthTable(trgVocabSize, 0);  // holds selected target words
+
+  // add firstNum most frequent words
+  for(WordIndex i = 0; i < firstNum_ && i < trgVocabSize; ++i)
+    trgTruthTable[i] = 1;
+
+  // collect unique words from source
+  // add aligned target words: mark trgTruthTable[word] to 1
+  for(auto word : srcBatch->data()) {
+    WordIndex srcIndex = word.toWordIndex();
+    if(shared_)
+      trgTruthTable[srcIndex] = 1;
+    // If srcIndex has not been encountered, add the corresponding target words
+    if (!srcTruthTable[srcIndex]) {
+      for (uint64_t j = wordToOffset_[srcIndex]; j < wordToOffset_[srcIndex+1]; j++)
+        trgTruthTable[shortLists_[j]] = 1;
+      srcTruthTable[srcIndex] = 1;
+    }
+  }
+
+  // Due to the 'multiple-of-eight' issue, the following O(N) patch is inserted
+  size_t trgTruthTableOnes = 0;   // counter for no. of selected target words
+  for (size_t i = 0; i < trgVocabSize; i++) {
+    if(trgTruthTable[i])
+      trgTruthTableOnes++;
+  }
+
+  // Ensure that the generated vocabulary items from a shortlist are a multiple-of-eight
+  // This is necessary until intgemm supports non-multiple-of-eight matrices.
+  for (size_t i = firstNum_; i < trgVocabSize && trgTruthTableOnes%8!=0; i++){
+    if (!trgTruthTable[i]){
+      trgTruthTable[i] = 1;
+      trgTruthTableOnes++;
+    }
+  }
+
+  // turn selected indices into vector and sort (Bucket sort: O(V))
+  std::vector<WordIndex> indices;
+  for (WordIndex i = 0; i < trgVocabSize; i++) {
+    if(trgTruthTable[i])
+      indices.push_back(i);
+  }
+
+  return New<Shortlist>(indices);
+}
+
+void BinaryShortlistGenerator::dump(const std::string& fileName) const {
+  ABORT_IF(mmapMem_.is_open(),"No need to dump again");
+  LOG(info, "[data] Saving binary shortlist dump to {}", fileName);
+  std::vector<char> blob = generateBlob();
+  saveBlobToFile(blob,fileName);
+}
+
+void BinaryShortlistGenerator::import(const std::string& filename, double threshold) {
```

Comment:
How fast does this run?  Can we kill the other shortlist?

Reply:
`import` is as slow as the original `LexicalShortlistGenerator`. As the biggest bottleneck is IO.

Reply:
I think by IO you mean C++ istreams are crazy slow at parsing things like floats, not necessarily raw byte copying speed.  Sigh, not something for you to fix, but an global issue with marian's adoption of istreams.